### PR TITLE
Update 2016-04-25-TF-Code-Structure.md

### PR DESCRIPTION
**Pull request by aytuncs**
mnist photo is 28 by 28 instead of 64 by 64

### CODE REVIEW

**Comment by knn940506** [comment]
## **텐서플로우란 ?**
  아마도 이 글을 읽는 분은 텐서플로우가 뭔지 이미 들어봤을겁니다.  텐서플로우는 구글에서 개발한 머신러닝 오픈소스 라이브러리입니다. 텐서플로우란 이름처럼 이 라이브러리는 텐서라는 다차원배열을 기본 데이터 타입으로 사용합니다. 학습 알고리즘은 물(tensor,데이터)이 수로(graph structure)를 통해 흘러가는(flow) 것처럼 실행됩니다.
  텐서플로우 코드들의 구조는 상당히 직관적이고 간단합니다. 학습을 위해 첫번째로 graph structure 을 정의 한 후에,  정의한 graph structure에 data stream 을 주입합니다. 
  시작하기전에 이 글이 아래 적힌 튜토리얼과 예제들을 참조하여 만들어졌다는 것을 알려드리고 싶습니다. 또한 여기 코드들은 실제 어플리케이션을 위한 텐서플로우 코드가 아니라 텐서플로우의 기본적인 흐름을 이해하기 쉽도록 쓰여진 것입니다. 그러니 더 실질적이고 개선된 코드를 위해 아래 예제 코드들을 참조하세요.

[Tutorials]
•	http://web.stanford.edu/class/cs224d/lectures/CS224d-Lecture7.pdf
•	http://www.slideshare.net/tw_dsconf/tensorflow-tutorial
[Example Codes]
•	https://github.com/nlintz/TensorFlow-Tutorials
•	https://github.com/aymericdamien/TensorFlow-Examples
•	https://github.com/pkmital/tensorflow_tutorials

## **텐서플로우 코드 흐름 백트랙킹**

**1. Graph structure 에 데이터 주입**
  텐서플로우 코드를 뒤에서 부터 검토해보겠습니다. 텐서플로우 코드의 마지막 단계는 무엇일까요 ? 결과값을 뽑아내기 위해, 준비한 데이터를 만들어 놓은graph structure(learning structure) 에 주입하는 것일 겁니다.

```
with tf.Session() as sess:
    tf.initialize_all_variables().run()
    sess.run(train_op, feed_dict={X: trX, Y: trY})
```

  `tf.Session()`은 `feed_dict`(데이터)를 `train_op`(learning structure) 에 집어넣음으로써 learning structure 을 구성합니다. 모든 변수들이 사용전에 초기화되어야함을 잊지 마세요. 
  여기서  Learning structure 은 연속적인 데이터의 흐름으로 여겨집니다. 만약 데이터 덩어리의  마지막 부분에서 한개의 데이터를 structure 에 입력하면, 자동적으로 데이터 덩어리 맨 앞까지 입력됩니다. 데이터 흐름이 연속적이니까요. 즉, `tf.Session()`이 필요로 하는것은 전체 graph structure이 아니라 structure 의 last output입니다. 코드에서 보면, last output 은`train_op`이고 `tf.Session()`이 access 합니다.

**2. 최적화**
  이어지는 질문은 `train_op`이 어디서 나오는가입니다. 이건 neural network 같은 learning structure 로 부터 학습되어져야 합니다. Neural network의 학습 과정은 network 를 모델을 주어진 조건에 일치시키기위한 optimizing weights가 전부입니다. 따라서 learning structure 의 마지막 단계는 optimization 입니다.` train_op`은 이런 최적화 과정을 통해 얻어집니다. 

  최적화 과정은 optimizer와 cost function 에 의해 이루어집니다. 

```
cost = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(y_pred, y_true))
train_op = tf.train.GradientDescentOptimizer(0.05).minimize(cost)
```

  이 코드에서`softmax_cross_entropy_with_logits`는 cost function 으로 사용되고, `GradientDescentOptimizer`는 optimizer로 사용됩니다. 여기와 이곳 링크에서 사용가능한 optimizers 와 cost functions 의 목록을 찾을수 있습니다. 물론 읽으시는 분의 목적에 맞춰서 고유 optimizer 이나 cost function 정의할 수 있습니다.

**3. Neural network structure**
  이제 우리는 최적화의 대상을 정의해야합니다. 어떤 structure 을 최적화해야 할까요 ? 답은 neural network structure 입니다. 따라서 우리는 최적화할 neural network structure 을 정의해야합니다. Neural network 의 기본 구조는 다음과 같습니다.

   1.	가중치 W , 편향 b 를 이용하여 input에 Linear transform(affine function) 을 적용합니다. 
           Ex ) y = Wx + b 
   2.	Learning structure 에 비선형성 부여를 위하여 y 에 activation function을 적용합니다.  
           Activation function 은 보통 rectified linear unit(ReLU)를 많이 사용합니다.
   3.	[0,1] 사이의 최종 결과값을 만들기 위해 softmax function 을 output layer의 마지막에 넣습니다.

```
# Step 1 : Affine function
W = tf.Variable(tf.random_normal([n_input, n_output]))
b = tf.Variable(tf.random_normal([n_input]))
y = tf.matmul(x, W) + b
# Step 2 : ReLU
y_pred = tf.nn.relu(y)
# Step 3 : Softmax (for the last layer)
softmax_out = tf.nn.softmax(y_pred)
```

  Step 3 는 cost function 이 `softmax_cross_entropy_with_logits function`처럼 이미 softmax function 을 포함하고 있을 경우 생략가능합니다. 
최종적으로,  training data 를 receiving 하는 노드들은 다음과 같이 정의됩니다.

 ```
   x = tf.placeholder("float", [None, n_input])
    y_true = tf.placeholder("float", [None, n_output])
```

 만약 목표가 28 * 28(784) 차원 MNIST 데이터를 이용해 0~9를 식별하는거라면n_input은 784가 되고n_output는 10을 입력하면 됩니다.

4.Summary
  요약하자면,  아래는 텐서플로우를 사용한 neural network 코드입니다.

```
import tensorflow as tf
import numpy as np

X = tf.placeholder("float", [None, n_input])
Y = tf.placeholder("float", [None, n_output])

W = tf.Variable(tf.random_normal([n_input, n_output]))
b = tf.Variable(tf.random_normal([n_input]))
y = tf.matmul(x, W) + b
y_pred = tf.nn.relu(y)

cost = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(y_pred, y_true))
train_op = tf.train.GradientDescentOptimizer(0.05).minimize(cost)

with tf.Session() as sess:
    tf.initialize_all_variables().run()
    sess.run(train_op, feed_dict={X: x_data, Y: y_data})
```

  이전에 언급했듯이, 이것은 완성된 코드가아닙니다. 코드의 개선을 위해 첫부분에 올린 예제 코드 링크들을 참고하도록 하세요. 실제 어플리케이션이선 neural network 학습 할때 더나은 결과를 위해서 수세대에 걸친 data minibatch(subset) 와 learning process(session)의 반복이 필요합니다. 또한 이미지 인식기술의 새로운 표준인 Convolutional neural network에서도 convolution layers 와 max pooling layers의 추가가 필요합니다.
  하지만 텐서플로우 코드의 기본적인 흐름은 동일합니다. Neural network의 structure 을 정의하고, 모델의 최적화를 위해 training data 를 structure 에 주입합니다. 그 후에 예측값과 실제값을 비교해서 평가를 합니다.
  이 글이 텐서플로우의 기본적인 이해에 도움이 되었으면 좋겠습니다. 부담갖지말고 페이스북이나 구글플러스에 피드백 올려주세요 !

제가 써본적이 없어서 제대로 해석했는지 모르겠네요... 실력자분들의 도움이 필요할거같습니다ㅠ
